apps/emqx_rule_engine/src/emqx_rule_funcs.erl

```diff
@@ -982,23 +982,35 @@ date_to_unix_ts(TimeUnit, Offset, FormatString, InputString) ->
     date_to_unix_ts(Unit, FormatString, InputString) - OffsetDelta.
 
 mongo_date() ->
-    erlang:timestamp().
+    maybe_isodate_format(erlang:system_time(millisecond), fun() -> erlang:timestamp() end).
 
-mongo_date(MillisecondsTimestamp) ->
-    convert_timestamp(MillisecondsTimestamp).
+mongo_date(MsTimestamp) ->
+    maybe_isodate_format(MsTimestamp, fun() -> convert_timestamp(MsTimestamp) end).
 
 mongo_date(Timestamp, Unit) ->
     InsertedTimeUnit = time_unit(Unit),
     ScaledEpoch = erlang:convert_time_unit(Timestamp, InsertedTimeUnit, millisecond),
-    convert_timestamp(ScaledEpoch).
+    mongo_date(ScaledEpoch).
 
-convert_timestamp(MillisecondsTimestamp) ->
-    MicroTimestamp = MillisecondsTimestamp * 1000,
+convert_timestamp(MsTimestamp) ->
+    MicroTimestamp = MsTimestamp * 1000,
     MegaSecs = MicroTimestamp div 1000_000_000_000,
     Secs = MicroTimestamp div 1000_000 - MegaSecs*1000_000,
     MicroSecs = MicroTimestamp rem 1000_000,
     {MegaSecs, Secs, MicroSecs}.
 
+maybe_isodate_format(MsTimestamp, Fun) ->
```

Comment:
This should work well and I don't think performance is a big issue here, but I think this could be made simpler if maybe_isodate_format took an Erlang timestamp (tuple with three elements as parameter). Then we could get rid of the fun as that code could be inlined directly in the calling function. If we are in the sqltest environment the timestamp tuple could be converted to and ISODate(...) string. It seems a bit backwards to optimize fot the test environment.